tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc

```diff
@@ -0,0 +1,573 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/c/experimental/filesystem/filesystem_interface.h"
+
+#include "google/cloud/storage/client.h"
+
+#include <string>
+#include <memory>
+#include <fstream>
+#include <iostream>
+
+#include <stdio.h>
+
+#include "tensorflow/c/tf_status.h"
+
+// Implementation of a filesystem for GCS environments.
+// This filesystem will support `gs://` URI scheme.
+#define TF_ReturnIfError(status) if(TF_GetCode(status) != TF_OK) return
+
+// We can cast `google::cloud::StatusCode` to `TF_Code` because they have the same integer values.
+// See https://github.com/googleapis/google-cloud-cpp/blob/6c09cbfa0160bc046e5509b4dd2ab4b872648b4a/google/cloud/status.h#L32-L52
+static inline void TF_SetStatusFromGCSStatus(const google::cloud::Status& gcs_status, TF_Status* status) {
+	TF_SetStatus(status, static_cast<TF_Code>(gcs_status.code()), gcs_status.message().c_str());
+}
+
+static void* plugin_memory_allocate(size_t size) { return calloc(1, size); }
+static void plugin_memory_free(void* ptr) { free(ptr); }
+
+static void ParseGCSPath(const char* fname, bool object_empty_ok, char** bucket, char** object, TF_Status* status) {
+  size_t scheme_index = strcspn(fname, "://");
+  char* scheme = (char*) malloc(scheme_index + 1);
+  sprintf(scheme, "%.*s", (int)scheme_index, fname);
```

Comment:
`snprintf` should be better than `sprintf` with `%.*` specifier.

However, since this does not cross plugin boundary, we should be able to use `std::string_view` to split the components and manipulate that via C++, as that is safer

Reply:
It means that I could use `std::string` to manipulate the path and `memcpy` to the `char*` later ? In fact, I generally avoid mixing `char*` and `std::string` as it may cause some pains

Reply:
`std::string_view` is a nice wrapper around both `std::string` and `char*` and has methods to keep track of both length and size so they don't get misused.

Reply:
I have updated `ParseGCSPath` to use `std::string_view`, about the schema, I will remove it when the changes with `TranslateName` is done

Reply:
I just have an idea. The current implementation of `CopyFile` for two filesystem different is creating a NewRandomAccessFile in the source filesystem, read it to a buffer, the append it to a NewAppendableFile in the the destination filesystem. Both `gcs` and `s3` will append to an temporary file. So instead doing like this, we can keep the `TranslateName` with the schema. When we pass to the cloud filesystem, it will do an Download if destination is local, Upload if source is local and CopyFile on cloud if both are cloud. The implementation can be modified as follow, if one path is cloud, use cloud, if both paths are local, use local.